Allow environment variables in #[files] attributes

hansl opened the pull request:
The environment variable must be passed to cargo test. Any variable name is valid and would be replaced. If a variable is not defined, the macro will error with the proper error message.

Fixes #276

hansl commented:
I changed the test to add a `HELLO_WORLD`. I'm going to add a proper test, but this is just to show an example.

Here is the result when running with the environment variable undeclared:

![CleanShot 2024-09-24 at 09 32 37@2x](https://github.com/user-attachments/assets/89cb532b-5c91-4134-9eb4-bee2b29051e3)

Here is the same test with a proper value (just having fun doing up and down on path):

![CleanShot 2024-09-24 at 09 33 20@2x](https://github.com/user-attachments/assets/809ebc8b-7fa5-418e-9243-dcf8b0a4390b)

Please note that if the environment variable is declared but doesn't go into the proper folder, it is equivalent to specifying the wrong path in `#[files]`. So the tests don't run.

la10736 reviewed (changes requested):
Really nice  JOB!!! Thanks for your contribution.

Can you please add some unit tests, a line in change log and add document this new feature in the https://github.com/la10736/rstest/blob/master/rstest/src/lib.rs#L954 and https://github.com/la10736/rstest/blob/a6a8cd9bc1441796a511694f9983243968f4bd51/README.md?plain=1#L233 ?

hansl commented on rstest_macros/src/parse/rstest/files.rs:
```diff
@@ -16,11 +16,54 @@ use crate::{
+            }
+        };
+
+        // Make sure cargo would rerun if the value of this environment variable changes.
+        println!("cargo::rerun-if-env-changed={var_name}");
```
It works in my project. Just rerunning `cargo test` without changing the environment variable will not rebuild the project, but changing it will rebuild.

hansl commented on rstest_macros/src/parse/rstest/files.rs:
```diff
@@ -213,10 +271,17 @@ impl VisitMut for ValueFilesExtractor {
                     .push(attr.error("Cannot use #[include_dot_files] more than once"))
             })
         }
+        let ignore_missing_env_vars = self.extract_ignore_missing_env_vars(node);
+        if !ignore_missing_env_vars.is_empty() {
+            ignore_missing_env_vars.iter().skip(1).for_each(|attr| {
+                self.errors
+                    .push(attr.error("Cannot use #[ignore_missing_env_vars] more than once"))
+            })
+        }
```
Done.

hansl commented on rstest_macros/src/parse/rstest/files.rs:
```diff
@@ -16,11 +16,54 @@ use crate::{
+/// Replace environment variables in a string.
+fn replace_env_vars(attr: &LitStrAttr, ignore_missing: bool) -> Result<String, syn::Error> {
+    let re =
+        Regex::new(r"\$([a-zA-Z_][a-zA-Z0-9_]*)|\$\{([^}]*)}").expect("Could not build the regex");
```
Let me add more of that.

hansl commented:
@la10736 Thanks for the review.

Added a new `#[base_dir = '...']` to allow to change the base dir for a test. It helps when using absolute paths (which cannot be used for `#[files(...)]`). Added tests for pretty much everything in this PR (checked locally with coverage). Also added a couple more integration tests.

PTAL

la10736 reviewed (approved):
That's an amazing work. Thanks again. 
I would have preferred that formatting was separated by code changing and some more unit tests about parsing: but I really appreciated your work and I should find a compromise with my desiderata :smile: 

I'm not sure if I really like the `base_dir` feature but maybe it can be useful in some context.

la10736 commented:
Just a note about tests and coverage: pass in every lines (100% coverage) doesn't means spot all  issues.... especially on on algorithms where the corner cases are not trivially identified by if branches. Anyway, your code and tests are really great!

la10736 commented on rstest_macros/src/parse/rstest/files.rs:
```diff
@@ -15,21 +9,114 @@ use crate::{
+                }
+            };
+
+            // Make sure cargo would rerun if the value of this environment variable changes.
+            println!("cargo::rerun-if-env-changed={var_name}");
```
Unfortunately it doesn't seem to work. I tried it this feature for base root and if just change, add or remove this variable the tests are not recompiled. If I add if to a build script it seams to work.

I'll remove it and add some documentation notes.

hansl replied:
I tested it on my side, let me investigate a bit.
